mm2src/lp_ordermatch.rs

```diff
@@ -494,41 +510,57 @@ struct GetOrderbookPubkeyItem {
     orders: PubkeyOrders,
 }
 
+/// Do not change this struct as it will break backward compatibility
+#[derive(Debug, Clone, Default, Deserialize, Serialize)]
+struct BaseRelProtocolInfo {
+    base: Vec<u8>,
+    rel: Vec<u8>,
+}
+
 #[derive(Debug, Deserialize, Serialize)]
 struct GetOrderbookRes {
     /// Asks and bids grouped by pubkey.
     pubkey_orders: HashMap<String, GetOrderbookPubkeyItem>,
+    #[serde(default)]
+    protocol_infos: HashMap<Uuid, BaseRelProtocolInfo>,
 }
 
-async fn process_get_orderbook_request(ctx: MmArc, base: String, rel: String) -> Result<Option<Vec<u8>>, String> {
-    fn get_pubkeys_orders(orderbook: &Orderbook, base: String, rel: String) -> (usize, HashMap<String, PubkeyOrders>) {
-        let asks = orderbook.unordered.get(&(base.clone(), rel.clone()));
-        let bids = orderbook.unordered.get(&(rel, base));
-
-        let asks_num = asks.map(|x| x.len()).unwrap_or(0);
-        let bids_num = bids.map(|x| x.len()).unwrap_or(0);
-        let total_orders_number = asks_num + bids_num;
-
-        // flatten Option(asks) and Option(bids) to avoid cloning
-        let orders = asks.iter().chain(bids.iter()).copied().flatten();
-
-        let mut uuids_by_pubkey = HashMap::new();
-        for uuid in orders {
-            let order = orderbook
-                .order_set
-                .get(uuid)
-                .expect("Orderbook::ordered contains an uuid that is not in Orderbook::order_set");
-            let uuids = uuids_by_pubkey.entry(order.pubkey.clone()).or_insert_with(Vec::new);
-            uuids.push((*uuid, order.clone()))
-        }
+fn get_pubkeys_orders(
+    orderbook: &Orderbook,
+    base: String,
+    rel: String,
+) -> (usize, HashMap<String, PubkeyOrders>, HashMap<Uuid, BaseRelProtocolInfo>) {
+    let mut protocol_infos = HashMap::new();
```

Comment:
maybe we should move this under `let mut uuids_by_pubkey = HashMap::new();` as `protocol_infos` is not needed before that

Reply:
Yes, it will be a bit more clear, I agree :slightly_smiling_face: I Will fix it shortly.